src/components/Navigation/Navigation.tsx

```diff
@@ -0,0 +1,262 @@
+import type { ReactNode, HTMLProps } from "react";
+import React, { isValidElement, useState } from "react";
+
+import classNames from "classnames";
+
+import NavigationLink from "./NavigationLink";
+import NavigationMenu from "./NavigationMenu";
+import type { GenerateLink, NavItem, NavMenu, LogoProps } from "./types";
+import { PropsWithSpread, SubComponentProps, Theme } from "types";
+import SearchBox, { SearchBoxProps } from "components/SearchBox";
+
+export type Props = PropsWithSpread<
+  {
+    /**
+     * By default the header is constrained to the width of the grid. Use this
+     * option to make the header take the full width of the page.
+     */
+    fullWidth?: boolean;
+    /**
+     * This function can be used to generate link elements when you don't want to
+     * use a standard HTML anchor.
+     */
+    generateLink?: GenerateLink | null;
+    /**
+     * The main navigation items that appear on the left hand side next to the logo.
+     */
+    items?: NavItem[] | null;
+    /**
+     * Additional navigation items that appear on the right hand side of the
+     * navigation banner.
+     */
+    itemsRight?: NavItem[] | null;
+    /**
+     * Additional props to be applied to the nav element that wraps the main
+     * navigation items on the left hand side.
+     */
+    leftNavProps?: HTMLProps<HTMLUListElement> | null;
+    /**
+     * The logo can be defined either by providing props for the standard logo
+     * or the full logo markup when a custom logo is needed.
+     */
+    logo: LogoProps | ReactNode;
+    /**
+     * Additional props to be applied to the nav element that wraps the
+     * left and right nav items.
+     */
+    navProps?: HTMLProps<HTMLElement> | null;
+    /**
+     * Additional props to be applied to the nav element that wraps the
+     * navigation items on the right hand side.
+     */
+    rightNavProps?: HTMLProps<HTMLUListElement> | null;
```

Comment:
I think the API of this component can be a little confusing (in particular `leftNavProps`, `navProps`, `rightNavProps`). Ideally props should be self-explanatory.

It seems implicit in the name that `navProps` belongs to the same category as `leftNavProps` and `rightNavProps` (I'd intuitively think it's for the navigation group in the middle), but that's not the case. `navProps` are the props for the `nav` element whilst `leftNavProps` and `rightNavProps` are for the props of the inner `ul`. 

I wonder if we need a separate set of props at all. Although I'm not a big fan of this pattern (prefer compound components), It might make more sense to colocate item groups attributes with items themselves (and add an optional className/props attribute to each item in the array.

This is a pattern that `react-components` users are already familiar with (it's how we handle this problem in the MainTable https://github.com/canonical-web-and-design/react-components/blob/main/src/components/MainTable/MainTable.tsx#L48-L57)

Reply:
Yeah it's not really brilliant is it. I guess I was hoping the comments in the props table would help to make it a bit more obvious. Could the names be clearer if they included "Element", like `navElementProps`?

I initially had a tuple for both the left and right navs, but realised that won't work for the case where you just need the right nav (like we need in MAAS in some cases).

Just to clarify, is your suggestion to have something like the following?

```jsx
<Navigation
    items={{
        items: [
            // The NavItem array 
        ],
        additionalLeftNavElementPropsHere: "",
    }}
    itemsRight={{
        items: [
            // The NavItem array 
        ],
        additionalRightNavElementPropsHere: "",
    }}
    navProps={{
        // We would still need this prop to get props to the nav element, the top level already spreads to <header>
        navElementPropsHere: "",
    }}
    additionalHeaderElementPropsHere=""
    ...
    />
```

Reply:
> Just to clarify, is your suggestion to have something like the following?

Something along these lines. I thought we could even get rid of explicitly setting named groups, and simply pass them in an array.

With this structure we get rid of the props naming problem entirely.

```
<Navigation
    items={[{
        items: [
            // The NavItem array 
        ],
        additionalLeftNavElementPropsHere: "",
    },
    {
        items: [
            // The NavItem array 
        ],
        additionalRightNavElementPropsHere: "",
    }}
    ...
/>
```

Reply:
> > Just to clarify, is your suggestion to have something like the following?
> 
> Something along these lines. I thought we could even get rid of explicitly setting named groups, and simply pass them in an array.
> 
> With this structure we get rid of the props naming problem entirely.
> 
> ```
> <Navigation
>     items={[{
>         items: [
>             // The NavItem array 
>         ],
>         additionalLeftNavElementPropsHere: "",
>     },
>     {
>         items: [
>             // The NavItem array 
>         ],
>         additionalRightNavElementPropsHere: "",
>     }}
>     ...
> />
> ```

I guess the awkward thing about using a tuple is that if you need to have items just on the right (like we sometimes do in MAAS) you have to include a null group `items={[null, {...}]}` which doesn't seem so obvious either...

Reply:
True, but (correct me if I'm wrong) that's only because we relying on an empty ul element to align items to the right. I couldn't find an example for such use case on vanilla, but ideally we should be able to pass a className (to the nav group or container) to achieve the desired layout instead?

Reply:
Even without having this use case covered by vanilla yet, simply adding `.u-flex--end` to `.p-navigation__nav` should do the trick? We would no longer need explicitly set item groups.

Reply:
Would that end up with a slightly awkward API in that there would be two mutually exclusive ways to align the groups? You'd have to either provide a tuple with both groups which would dictate which group was right or left (so you couldn't also have a class/prop to align them) or in the case that you only want to display one group you'd include one item in the array with a possible class/prop to align to the right.

In cases like MAAS where we display both states at different times we'd then have to do a bit of work to include the right things in the array with the right classes/props. That's certainly not hard to do, but I wonder if that becomes more messy to work with than isolating the props for left and right.

Reply:
@huwshimi @petermakowski it seems this PR leans into being a discussion about how we approach props API in our components. The more complex the component gets (like `MainTable`, this navigation, etc) the harder it is to pass all needed data as props and maintain flexibility that we are used to with Vanilla (being able to adjust HTML structure, class names, etc).

I'm starting to think that the way MUI approaches it (by using the composable child components) seems to be more flexible in this regard (for example: https://mui.com/material-ui/react-bottom-navigation/). 

I feel that we had a similar discussion in the past (when Modular Table was started maybe?), but I don't remember where it landed. 

I don't know what's the best approach, but the more complicated "items" props will be the harder it will be to use it or maintain in future. 
Or maybe some compromise would be to have a set of smaller composable components (`NavigationBar`, `NavigationItems`, `NavigationSearch`, etc) that you can use to custom build any "non standard" configuration for navigation, but we also provide `MainNavigation` component that you just pass props to that should cover most cases - kind of similar to what we have with `MainTable` being a controller component for  Table/TableRow/TableCell, etc…

Reply:
Agreed @bartaz 

I'm in favour of breaking down into smaller composable components. There are a few different patterns that could work, I already mentioned [compound components](https://kentcdodds.com/blog/compound-components-with-react-hooks), simply grouping subcomponents by using a single prefix where sharing state is not required could work as well. 

The mui example you posted could work, too (although not everything I agree with, e.g. their approach of using boolean props to differentiate subcomponents probably isn't something we should follow - `<Grid container>
  <Grid item /> (...) </Grid>` ). https://mui.com/material-ui/react-grid/

Reply:
We we initially talked about this we looked at the pattern of Material and other libraries and they often provided components without complex interactions, which leaves implementing the interactions/behaviour/functionality up to the consumers of the lib.

We however decided that react-components should provide both visual and functionality as, unlike a component library which is trying to be flexible, we want consistent behaviour across our sites/apps.

It's possible that we could revisit how we implement this, but we should make sure that we're able to implement behaviours in this lib.

Reply:
Okay, let's revisit this later. I created a separate issue to track this: https://github.com/canonical-web-and-design/react-components/issues/774